api/envoy/extensions/rbac/matchers/upstream_port/v3/upstream_port_matcher.proto

```diff
@@ -0,0 +1,29 @@
+syntax = "proto3";
+
+package envoy.extensions.rbac.matchers.upstream_port.v3;
+
+import "envoy/type/v3/range.proto";
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.rbac.matchers.upstream_port.v3";
+option java_outer_classname = "UpstreamPortMatcherProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: RBAC upstream port range matcher plugin]
+// [#extension: envoy.rbac.matchers.upstream_port]
+
+// This is configuration for matching upstream port.
+message UpstreamPortMatcher {
+  // A port range that will be used to match the upstream port.
+  // This matcher requires a filter in the chain to have saved the upstream address(es) in the
+  // filter state before the matcher is executed by RBAC filter. The state should be saved with key
+  // `envoy.stream.upstream_addresses` (See
+  // :repo:`upstream_address_set.h<source/common/stream_info/upstream_address_set.h>`).
+  // Also, See :repo:`proxy_filter.cc<
+  // source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc>` for an example of a
+  // filter which populates the FilterState. If the key is not found in the filter state, the
+  // matcher will evaluate to `false`.
```

Comment:
Can you explain how this works with the set semantics? I like the explanation for upstream IP addresses that you now have, but I'm a bit confused here actually. How would IP and port, in separate matchers, sensibly combine with ANY semantics? 

Let's say my bag of upstream addresses is {10.0.0.1:80, 10.0.0.2:1234}. Can I write a policy that matches on a specific IP:port in the bag with these matchers?

Or is the assumption the port is always the same (likely for dynamic forward, but maybe not for other situations, or if SRV records are used).

Reply:
OK, let's say my upstream addresses are now {10.0.0.1:1234, 10.0.0.2:80} in the next request. The matcher for 10.0.0.1:80 will still match this set of addresses.

Reply:
Thanks @htuch .

I think we should be able to compose the expression like: `ALLOW` IFF  { {(`ip=10.0.0.1`) `AND` (`port=80)` } `OR ` {(`ip=10.0.0.2`) AND (`port=1234`) }

This way we only ALLOW a set of ip:port combinations.  There is still an issue when the bag has more IPs than the rules.
For example, if the bag contains 3 items and we only have OR rules for 2, then the matcher OR rules will still be not be able to cover all.

This complexity could be avoided by adding the `match_type` attribute as I mentioned above. But for the case of DFP, we need not worry about them i think.

We could document the above gap . When we add generic ip address bag , we need to implement the `match_type`.

Reply:
Can we just have a matcher that does both IP CIDR range and port? This seems simplest.

Reply:
I think by combining IP CIDR and PORT in the same matcher, it will reduce composability. We should be able to achieve the same by composing IP matcher and PORT matcher , no?

Reply:
@mattklein123 @alyssawilk @lizan after some offline discussion with @conqerAtapple, we reached the conclusion that it might be better to have the filter state be a single address (which matches what we are actually doing today in DFP). If we keep a set of addresses, the ANY semantics gets weird in a bunch of cases, and it's unclear if you can write sensible matchers if you don't actually know the IP:port that is going to be used on the wire.

So, @conqerAtapple is fine with switching from set to singleton. Does anyone have any objections to this?

Reply:
Thanks @htuch . To add more context, the idea of  IP `set` was for future upstream proxies which might be populating the set with not just DNS resolved IPs but maybe LB subsets. Then RBAC can determine if a subset is ALLOW'ed. But in the absence of upstream filters, we currently dont have any use for the `set`. So I am ok with replacing set with a single IP.

Reply:
I think single IP makes sense for now, but I think for this to be useful you need a concept of optional port / wildcard port / port range? Can we make sure to work that in?